appserver/microprofile/health/src/main/java/org/glassfish/microprofile/health/HealthReporter.java

```diff
@@ -0,0 +1,148 @@
+        } catch (RuntimeException e) {
+            LOGGER.log(Level.SEVERE, "Health check failed", e);
+            return buildHealthCheckResponse(healthCheck.getClass().getName(), e);
+        }
+    }
+
+    private static HealthCheckResponse buildHealthCheckResponse(String name, Exception e) {
+        return HealthCheckResponse.builder()
+                .down()
+                .name(name)
+                .withData("rootCause", e.getMessage())
+                .build();
+    }
+
+    public enum ReportKind {
+        LIVE,
+        READY,
+        STARTED,
+        ALL;
+
+        private Stream<HealthCheck> healthChecks() {
+            Instance<HealthCheck> select = CDI.current().select(HealthCheck.class);
+            return switch (this) {
+                case LIVE -> select.select(Liveness.Literal.INSTANCE).stream();
+                case READY -> select.select(Readiness.Literal.INSTANCE).stream();
+                case STARTED -> select.select(Startup.Literal.INSTANCE).stream();
+                case ALL -> Stream.of(
+                        select.select(Liveness.Literal.INSTANCE).stream(),
+                        select.select(Readiness.Literal.INSTANCE).stream(),
+                        select.select(Startup.Literal.INSTANCE).stream()
+                ).flatMap(s -> s);
+            };
+        }
+
+        private HealthCheckResponse.Status getEmptyResponse() {
+            return switch (this) {
+                case LIVE -> getValue(MP_DEFAULT_STARTUP_EMPTY_RESPONSE)
+                        .orElse(HealthCheckResponse.Status.UP);
+                case READY -> getValue(MP_DEFAULT_READINESS_EMPTY_RESPONSE)
+                        .orElse(HealthCheckResponse.Status.UP);
+                case STARTED, ALL -> HealthCheckResponse.Status.UP;
+            };
+        }
+    }
+
+    public HealthReport getReport(ReportKind reportKind) {
+        HealthCheckResponse.Status emptyResponse = HealthCheckResponse.Status.UP;
+
+        Stream<HealthCheck> healthChecks = applicationHealthChecks.values()
+                .stream()
+                .flatMap(Collection::stream);
+//                Stream.concat(reportKind.healthChecks(), applicationHealthChecks.stream());
+
+        List<HealthCheckResponse> healthCheckResults = healthChecks
+                .map(HealthReporter::callHealthCheck)
+                .toList();
+
+        HealthCheckResponse.Status overallStatus;
+        if (healthCheckResults.isEmpty()) {
+            overallStatus = emptyResponse;
+        } else {
+            overallStatus = healthCheckResults.stream()
+                    .map(HealthCheckResponse::getStatus)
+                    .filter(HealthCheckResponse.Status.DOWN::equals)
+                    .findFirst()
+                    .orElse(HealthCheckResponse.Status.UP);
+        }
+        return new HealthReport(overallStatus, healthCheckResults);
+
+    }
+
+
+    public void addHealthCheck(String contextPath, HealthCheck healthCheck) {
```

Comment:
I would rename `contextPath` to `contextName`. It has nothing to do with a "path", it's just a context to group several checks. In GlassFish, it would be application name.